### ISSUE TITLE
Check if SEQAN_CONSTEXPR_IF can be used to speed up the alignment code 

### ISSUE THREAD

**Issue by rrahn** [seqan/seqan#2120]
We can add a macro `SEQAN_CONSTEXPR_IF` in if statements to enable the `if constexpr` feature if c++17 is enabled, which can `no-op` some branches based on compile time decisions.
Before c++17, the compiler might or might not decide to remove a branch.
This also effects some portability issues, because the c++17 based `if constexpr` is more powerful than the current behaviour, as all branches (also the potentially removed one) must be executable for all types.

The following code can only be compiled if c++17 is enabled:
```cpp
#if __cplusplus >= 201703L
#define SEQAN_CONSTEXPR_IF constexpr
#else
#define SEQAN_CONSTEXPR_IF
#endif

struct foo
{
    int m;
};

struct bar 
{
    int n;
};    

template <typename Type>
void test(Type const foobar)
{
    if SEQAN_CONSTEXPR_IF (std::is_same<Type, foo>::value)
    {
        std::cout << foobar.m << std::endl;
    }
    else
    {
        std::cout << foobar.n << std::endl;
    }
}

int main()
{
    test(bar{});
}

```

**Comment by h-2** [comment]
Yes, this will be only a replacement for regular if clauses, not type-specific branching. Since we test with c++14 by default, we don't run the risk of mixing these up, I think.

But, we would need to add a c++17-ified compiler to travis.

**Comment by h-2** [comment]
Btw, I would propose
```cpp
#if __cplusplus >= 201703L
#define SEQAN_IF_CONSTEXPR if constexpr
#else
#define SEQAN_IF_CONSTEXPR if
#endif
```
i.e. if_constexpr instead of constexpr_if and including the if in the macro.
But no strong feelings about it.

**Comment by h-2** [comment]
I have added a `g++7-stdc++17` build to the freebsd nightlies that we should be getting as of tonight.

**Comment by h-2** [comment]
I'll take this since you have quite the issue list still.

**Comment by h-2** [comment]
implemented via #2211
